sling/file/recordio.cc

```diff
@@ -0,0 +1,449 @@
+  // Allocate output buffer.
+  output_.resize(options.buffer_size);
+  position_ = 0;
+
+  // Write file header.
+  memset(&info_, 0, sizeof(info_));
+  info_.magic = MAGIC;
+  info_.hdrlen = sizeof(info_);
+  info_.compression = options.compression;
+  info_.chunk_size = options.chunk_size;
+  memcpy(output_.end(), &info_, sizeof(info_));
+  output_.appended(sizeof(info_));
+  position_ += sizeof(info_);
+}
+
+RecordWriter::RecordWriter(const string &filename,
+                           const RecordFileOptions &options)
+    : RecordWriter(File::OpenOrDie(filename, "w"), options) {}
+
+RecordWriter::RecordWriter(File *file)
+    : RecordWriter(file, default_options) {}
+
+RecordWriter::RecordWriter(const string &filename)
+    : RecordWriter(filename, default_options) {}
+
+RecordWriter::~RecordWriter() {
+  CHECK(Close());
+}
+
+Status RecordWriter::Close() {
+  // Check if file has already been closed.
+  if (file_ == nullptr) return Status::OK;
+
+  // Flush output buffer.
+  Status s = Flush();
+  if (!s.ok()) return s;
+
+  // Close output file.
+  s = file_->Close();
+  file_ = nullptr;
+
+  return s;
+}
+
+Status RecordWriter::Flush() {
+  if (output_.empty()) return Status::OK;
+  Status s = file_->Write(output_.begin(), output_.size());
+  if (!s.ok()) return s;
+  output_.clear();
+  return Status::OK;
+}
+
+Status RecordWriter::Write(const Record &record) {
+  // Compress record value if requested.
+  Slice value;
+  if (info_.compression == SNAPPY) {
+    // Compress record value.
+    SliceSource source(record.value);
+    compressed_data_.clear();
+    snappy::Compress(&source, &compressed_data_);
+    value = Slice(compressed_data_.begin(), compressed_data_.end());
+  } else if (info_.compression == UNCOMPRESSED) {
+    // Store uncompressed record value.
+    value = record.value;
+  } else {
+    return Status(1, "Unknown compression type");
+  }
+
+  // Compute on-disk record size estimate.
+  size_t maxsize = MAX_HEADER_LEN + record.key.size() + value.size();
+
+  // Records cannot be bigger than the chunk size.
+  size_t size_with_skip = maxsize + MAX_SKIP_LEN;
+  CHECK_LE(size_with_skip, info_.chunk_size)
+      << "Record too big (" << size_with_skip << " bytes), "
+      << "maximum is " << info_.chunk_size << " bytes";
+
+  // Flush output buffer if it does not have room for record.
+  if (maxsize > output_.remaining()) {
+    Status s = Flush();
+    if (!s.ok()) return s;
+  }
+
+  // Check if record will cross chunk boundary.
+  if (info_.chunk_size != 0) {
+    uint64 chunk_used = position_ % info_.chunk_size;
+    if (chunk_used + size_with_skip > info_.chunk_size) {
+      // Write filler record. For a filler record, the record size includes
+      // the header.
+      Header filler;
+      filler.record_type = FILLER_RECORD;
+      filler.record_size = info_.chunk_size - chunk_used;
+      filler.key_size = 0;
+      int hdrlen = WriteHeader(filler, output_.end());
+      output_.appended(hdrlen);
```

Comment:
shouldn't there also be a position_ += hdrlen here?

Reply:
This is somewhat subtle, but as stated in the comment above, the record size for a filler record includes the header.